File: gpt-4-turbo/stage2/acc_deviceptr.f90
! Assuming acc_testsuite.Fh provides necessary infrastructure for the test suite
!T1:acc_deviceptr,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        USE, INTRINSIC :: ISO_C_BINDING, ONLY : C_PTR, C_F_POINTER
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER, DIMENSION(:), ALLOCATABLE :: host_array
        TYPE(C_PTR) :: device_ptr
        LOGICAL :: isOnDevice
        INTEGER :: errors = 0
        INTEGER :: i, N

        ! Initialize a test array
        N = 100
        ALLOCATE(host_array(N))
        host_array = (/(i, i=1,N)/)

        ! Use OpenACC to manage the array on the device
        !$acc enter data copyin(host_array(:))

        ! Get the device pointer
        device_ptr = acc_deviceptr(host_array)

        ! Check if the device_ptr is not null which means the data is present on the device
        IF (.NOT. C_ASSOCIATED(device_ptr)) THEN
          errors = errors + 1
        END IF

        !$acc exit data delete(host_array(:))

        ! Final assessment
        IF (errors .eq. 0) THEN
          test1 = .FALSE.   ! No errors
        ELSE
          test1 = .TRUE.    ! Errors occurred
        END IF

        ! Clean up
        DEALLOCATE(host_array)
      END

      PROGRAM main
        USE OPENACC
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
        LOGICAL :: test1
        failed = .FALSE.
        failcode = 0
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
        ! Use standard Fortran method of stopping the program to ensure compatibility
        STOP failcode
      END PROGRAM main

File: gpt-4-turbo/stage2/acc_deviceptr.cpp
#include <iostream>
#include <stdlib.h>
#include <openacc.h>

#ifndef SEED
#define SEED 123
#endif

#ifndef NUM_TEST_CALLS
#define NUM_TEST_CALLS 10
#endif

// Test for acc_deviceptr
// T1: acc_deviceptr, V:2.7-3.3
int test_acc_deviceptr(){
    int err = 0;
    srand(SEED);

    const int size = 1024;
    float* host_array = new float[size];
    // Initialize host array
    for(int i = 0; i < size; ++i){
        host_array[i] = static_cast<float>(rand()) / RAND_MAX;
    }

    // Move data to device - ensures there is a device side copy for acc_deviceptr to find
    #pragma acc enter data copyin(host_array[0:size])

    // Fetch device pointer
    float* device_ptr = (float*) acc_deviceptr(host_array);
    if (device_ptr == NULL) {
        std::cerr << "acc_deviceptr returned NULL, indicating data is not present on the device." << std::endl;
        err++;
    } else if (device_ptr == host_array) {
        std::cerr << "acc_deviceptr returned host address instead of device address." << std::endl;
        err++;
    }

    // Cleanup
    #pragma acc exit data delete(host_array[0:size])
    delete[] host_array;

    return err;
}

int main(){
    int failcode = 0;
    int failed = 0;
    
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed += test_acc_deviceptr();
    }

    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
    
    return failcode;
}